### PR TITLE
Pressure solver for vertically stretched grids

### LINKED ISSUE #46: Poisson solver that allows for variable vertical grid spacing.
@christophernhill has the details figured out. Sounds like it just needs to be implemented.

### PR DESCRIPTION
**Pull request by ali-ramadhan**
This PR aims to introduce and integrate Poisson solvers for vertically stretched grids developed by @christophernhill and @maleadt. Once this is merged alongside with PR #283 we should be able to support vertically stretched grid.

For PPN boundary conditions, the system is solved for by eigenvalue-eigenvector factorization via FFTs in the horizontal and we end up with `Nx*Ny` tridiagonal systems, one for each column so they can be solved in parallel. We might be able to speed up the already fast tridiagonal solve via cyclic reduction. For PNN boundary conditions, it should work if we use an FFT in the x and a DCT in the y.

So far I have a working 1D solver. The two tricks that had to be made were:
1. @christophernhill figured out that we need to stop the algorithm once a zero eigenvalue is reached. This should only occur at the last element for our tridiagonal systems.
2. There are `Nz+1` faces in the vertical but we only keep track of the first `Nz`. The tridiagonal system needs to explicitly account for all `Nz+1` faces. Thus we impose the staggered Neumann boundary conditions at `k=Nz+1` and the (Nz+1)th entry of right-hand-side is zero. (Chris probably told me this but I forgot...) Previously, the DCT did this for us as all the eigenfunctions satisfied the boundary condition anyways.

I tested it against our existing solver but it doesn't work for variable dz yet. Not sure if it's because of my implementation of the Laplacian in PR #283 or because I discretized the vertical derivative incorrectly here. Just opening this PR early so everyone's in the loop.

I'm not 100% sure whether we will be able to utilize cyclic reduction:
* I think you need a symmetric tridiagonal system (or at least block-symmetric) which we don't have when the grid is vertically stretched. Maybe this is a sign that I discretized things wrong?
* It can be modified for Neumann boundary conditions (which break the symmetry on the first and last rows) but in this case it seems cyclic reduction must be applied recursively `m+1` times for a system of size `Nz = 2^m + 1`. The optimal number of cyclic reductions is `log2(log2(Nz))` so it's possible that we don't gain much. There may also be numerical accuracy issues with taking too many cyclic reduction steps.

Some thoughts:
* The batched cyclic reduction solver @maleadt developed only works for systems sizes that are powers of 2. As we need to solve a system of `Nz+1` equations, this means models with `2^m - 1` vertical levels will be the most efficient. If we can use it here, then we might need a non-cyclic reduction solver for models without `2^m - 1` vertical levels.
* Probably not within this scope of this PR, but this solver should work on CPUs and GPUs and will eliminate the need for two different solvers, so we can probably unify solvers. We may still want to use cyclic reduction wherever we can. We should be able to use it for uniform vertical spacing and `Nz = 2^m - 1`. @maleadt's implementation was ~3x faster than the cuSPARSE implementation and ~22x faster than the serial implementation, so it would be a huge speedup.

cc @sandreza as I might ask you questions about this

This work is based off:
https://gist.github.com/christophernhill/8952075d6e95df28adf36f5dff64115a
https://gist.github.com/christophernhill/a43317694c6d65ecd27758c1bdcdf3f1
https://gist.github.com/maleadt/1ec91b3b12ede9898958c95596cabe8b

Resolves #46
Helps with #47
Helps with #211

### CODE REVIEW

**Comment by ali-ramadhan** [comment]
> I think you need a symmetric tridiagonal system (or at least block-symmetric) which we don't have when the grid is vertically stretched. Maybe this is a sign that I discretized things wrong?

After talking to @christophernhill and @jm-c indeed I discretized the vertical derivative incorrectly. Accounting for dzF and dzC, you get 1/dzF factors on the diagonals and 1/dzC on the sub- and super-diagonal and it comes out symmetric.

**Comment by ali-ramadhan** [comment]
This PR now adds a test to ensure that a Poisson equation with arbitrary/random source term can be solved on a vertically stretched grid with the `BatchedTridiagonalSolver` up to machine precision with the staggered grid difference operators.

I tried to add a struct for a vertically stretched Poisson solver but it required a finalized vertically stretched grid (currently under development in PR #543) so I think it'll be better if they're developed together.

The test will then be refactored to make use of the `VerticallyStretchedCartesianGrid`.

**Comment by glwagner** [comment]
> There are Nz+1 faces in the vertical but we only keep track of the first Nz. The tridiagonal system needs to explicitly account for all Nz+1 faces. Thus we impose the staggered Neumann boundary conditions at k=Nz+1 and the (Nz+1)th entry of right-hand-side is zero. (Chris probably told me this but I forgot...) Previously, the DCT did this for us as all the eigenfunctions satisfied the boundary condition anyways.

This provides another reason to include all the `Nz+1` face data for face-centered fields. Will this feature be easy to implement in the future?

File: test/test_solvers.jl [hunk 1]
```diff
@@ -296,22 +297,169 @@ function can_solve_batched_tridiagonal_system_with_3D_functions(arch, Nx, Ny, Nz
     return Array(ϕ) ≈ ϕ_correct
 end
 
+function vertically_stretched_grid_div_free_velocity(arch, Nx, Ny, zF)
```
Comment (glwagner) [inline test/test_solvers.jl]: It looks like this test solves the Poisson equation with an eigenfunction expansion. Is this correct? I don't think it is a literal test of the enforcement of a divergence free velocity field via the fractional step algorithm. If so, I think the name of this test should be renamed. Also, much of the complexity associated with computing the divergence of some velocity field, etc can be avoided. I think we can just randomly construct `R`, compute the solution `ϕ` to the Poisson equation `∇²ϕ = R`, and then compare `R` to `∇²ϕ`. Right?

Comment (glwagner) [reply test/test_solvers.jl]: This test could also use a docstring I think because it is quite complicated --- may be hard to follow for future developers. I know many of our tests don't have docstrings, but maybe we should change that?

**Comment by ali-ramadhan** [comment]
> This provides another reason to include all the `Nz+1` face data for face-centered fields. Will this feature be easy to implement in the future?

So that comment is actually outdated now. The Neumann boundary conditions can be built into the `Nz x Nz` tridiagonal system.

> Pretty simple PR I think because it basically just adds a test (please confirm... ?)

Yes. It was originally going to add a vertically stretched pressure solver but then I realized it would be much better code design to do the refactoring I did PR #589 first then add the solver in.

So this PR just adds a test that shows you can implement a vertically stretched pressure solver using the `BatchedTridiagonalSolver`.

> I think the test may be incorrectly named and can also be greatly simplified. I'm not sure if the test is 1) not achieving its intended purpose or 2) is over-complicated. Either way I think it should be cleaned up a bit before merging.

Yes it is unfortunately over-complicated because there is no abstraction for a vertically stretched grid or a stretched pressure so they are implemented in the test itself. This test will be massively cleaned up in PR #543.

File: test/test_solvers.jl [hunk 2]
```diff
@@ -296,22 +297,169 @@ function can_solve_batched_tridiagonal_system_with_3D_functions(arch, Nx, Ny, Nz
...
+    @inline ∂x²(i, j, k, Δx, f)       = (∂x_caa(i, j, k, Δx, f)  - ∂x_caa(i-1, j, k, Δx, f))  / Δx
+    @inline ∂y²(i, j, k, Δy, f)       = (∂y_aca(i, j, k, Δy, f)  - ∂y_aca(i, j-1, k, Δy, f))  / Δy
+    @inline ∂z²(i, j, k, ΔzF, ΔzC, f) = (∂z_aac(i, j, k, ΔzF, f) - ∂z_aac(i, j, k-1, ΔzF, f)) / ΔzC[k]
+
+    @inline div_ccc(i, j, k, Δx, Δy, ΔzF, u, v, w) = ∂x_caa(i, j, k, Δx, u) + ∂y_aca(i, j, k, Δy, v) + ∂z_aac(i, j, k, ΔzF, w)
+
+    @inline ∇²(i, j, k, Δx, Δy, ΔzF, ΔzC, f) = ∂x²(i, j, k, Δx, f) + ∂y²(i, j, k, Δy, f) + ∂z²(i, j, k, ΔzF, ΔzC, f)
+
+    #####
+    ##### Generate "fake" vertically stretched grid
```
[the remark this answers is not in the document]
Comment (ali-ramadhan) [reply test/test_solvers.jl]: It's just fake because I'm actually trying to solve for pressure on a vertically stretched grid but create and pass a `RegularCartesianGrid` to the solver.

File: src/Solvers/batched_tridiagonal_solver.jl [hunk 3]
```diff
@@ -39,9 +39,9 @@ where `dl` stores the lower diagonal coefficients `a(i, j, k)`, `d` stores the d
 
 `params` is an optional named tuple of parameters that is accessible to functions.
 """
-function BatchedTridiagonalSolver(arch=CPU(); dl, d, du, f, grid, params=nothing)
+function BatchedTridiagonalSolver(arch=CPU(), FT=Float64; dl, d, du, f, grid, params=nothing)
```
[the remark this answers is not in the document]
Comment (ali-ramadhan) [reply src/Solvers/batched_tridiagonal_solver.jl]: I wanted to follow the same names as https://docs.julialang.org/en/v1/stdlib/LinearAlgebra/#LinearAlgebra.Tridiagonal for consistency, but we can change them. Would it be okay if we changed the names in a different PR like #589 to keep this focused on adding the test?

File: test/test_solvers.jl (hunk 1 above)
Comment (ali-ramadhan) [reply test/test_solvers.jl]: > It looks like this test solves the Poisson equation with an eigenfunction expansion. Is this correct? I don't think it is a literal test of the enforcement of a divergence free velocity field via the fractional step algorithm. If so, I think the name of this test should be renamed.

That's a good point. I renamed the test to `vertically_stretched_poisson_solver_correct_answer`.

> Also, much of the complexity associated with computing the divergence of some velocity field, etc can be avoided. I think we can just randomly construct `R`, compute the solution `ϕ` to the Poisson equation `∇²ϕ = R`, and then compare `R` to `∇²ϕ`. Right?

That's exactly what the test used to do but I think the right-hand side needs to actually satisfy the boundary condition $G_u \dot \hat{n} = 0$ according to Marshall, Hill, Perelman, & Adcroft (section 3.2, 1997).

If I just give it a random right hand side with zero mean the test fails. I believe that the discrete cosine transform was enforcing this boundary condition for us but now we actually have to enforce us ourselves, but not sure how to show this.

Comment (glwagner) [reply test/test_solvers.jl]: > If I just give it a random right hand side with zero mean the test fails.

Erm... `R` doesn't have boundary values... ? I don't think its true that `R` needs to satisfy some criterion for the test to be valid. You have have to fill the halos for `ϕ` before computing `∇²ϕ`... ? Or, perhaps you can circumvent the issue of calculating the Laplacian across the boundary by only comparing `∇²ϕ` to `R` on interior points between 2:Nx-1... ?

Comment (glwagner) [reply test/test_solvers.jl]: > That's a good point. I renamed the test to vertically_stretched_poisson_solver_correct_answer.

Ok, but isn't this actually a test of the *tridiagonal* solver on a uniform grid?

File: src/Solvers/batched_tridiagonal_solver.jl (hunk 3 above)
[the remark this answers is not in the document]
Comment (glwagner) [reply src/Solvers/batched_tridiagonal_solver.jl]: Oh. That's fine. But maybe the docstring should use `dl` `d` and `du` then. Anyways, I'm fine if this is addressed in the future.

File: test/test_solvers.jl (hunk 1 above)
Comment (ali-ramadhan) [reply test/test_solvers.jl]: I am filling in the halos properly. I don't think we should shortcut the test and not check the boundaries, getting the correct solution at the boundary is an important part of the test.

Using a completely random right-hand-side with zero mean, you end up with the correct solution everywhere except at the last vertical grid point where there is a systematic error that is the same at all the columns, so I feel like there may be a simple bug somewhere that is somehow not present when $G_u \dot \hat{n} = 0$.

**Comment by ali-ramadhan** [comment]
Actually I will wait until PRs #543 and #589 are merged and use this PR to add a proper solver.
